src/downloader.ts

```diff
@@ -0,0 +1,194 @@
+import { chmodSync } from "node:fs";
+import { Octokit } from "octokit";
+import { fetch, ofetch } from "ofetch";
+import { coerce, rcompare } from "semver";
+import {
+	ExtensionContext,
+	ProgressLocation,
+	Uri,
+	commands,
+	window,
+	workspace,
+} from "vscode";
+import { Commands } from "./commands";
+
+export const selectAndDownload = async (
+	context: ExtensionContext,
+): Promise<string | undefined> => {
+	const versions = await window.withProgress(
+		{
+			location: ProgressLocation.Notification,
+			title: "Fetching Biome versions",
+			cancellable: false,
+		},
+		async () => {
+			return await getVersions(context);
+		},
+	);
+
+	const version = await askVersion(versions);
+
+	if (!version) {
+		return undefined;
+	}
+
+	return await window.withProgress(
+		{
+			location: ProgressLocation.Notification,
+			title: `Downloading Biome ${version}`,
+			cancellable: false,
+		},
+		async () => {
+			await commands.executeCommand(Commands.StopServer);
+			await download(version, context);
+			await commands.executeCommand(Commands.RestartLspServer);
+			return version;
+		},
+	);
+};
+
+export const updateToLatest = async (context: ExtensionContext) => {
+	await window.withProgress(
+		{
+			location: ProgressLocation.Notification,
+			title: "Updating Biome version",
+			cancellable: false,
+		},
+		async () => {
+			const versions = await getVersions(context);
+			const version = versions[0];
+			await commands.executeCommand(Commands.StopServer);
+			await download(version, context);
+			await commands.executeCommand(Commands.RestartLspServer);
+		},
+	);
+};
+
+/**
+ * Download the Biome CLI from GitHub
+ *
+ * @param version The version to download
+ */
+const download = async (version: string, context: ExtensionContext) => {
+	// Find the correct asset to download
+	const octokit = new Octokit({
+		request: {
+			fetch: fetch,
+		},
+	});
+
+	const release = await octokit.request(
+		"GET /repos/{owner}/{repo}/releases/tags/{tag}",
+		{
+			owner: "biomejs",
+			repo: "biome",
+			tag: `cli/v${version}`,
+		},
+	);
+
+	const platformArch = `${process.platform}-${process.arch}`;
+
+	// Find the asset for the current platform
+	const asset = release.data.assets.find(
+		(asset) =>
+			asset.name ===
+			`biome-${platformArch}${process.platform === "win32" ? ".exe" : ""}`,
+	);
+
+	if (!asset) {
+		window.showErrorMessage(
+			`The specified version is not available for your platform/architecture (${platformArch}).`,
+		);
+	}
+
+	const bin = await ofetch(asset.browser_download_url, {
```

Comment:
Can we use [`undifici`](https://www.npmjs.com/package/undici) instead of `ofetch`? Node.js 18 uses `undici` under the hoods

Reply:
Yes, I supposed we could. I'd like to know why undici is preferred, though.

Reply:
Because it's the library that Node.js 18 uses for fetch, so we'll have zero surprises once we drop v16. We used undici in Astro as a polyfill for a long time and it's production ready.

Reply:
Removed ofetch in favor of undici.